src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,241 @@
+    const float cutoffHz = 0.01f * positionConfig()->altitude_d_lpf; // default 1Hz, time constant about 160ms
+    const float gain = pt2FilterGain(cutoffHz, taskIntervalSeconds);
+    pt2FilterInit(&altHoldDeltaLpf, gain); 
+
+    altHoldState.isAltHoldActive = false;
+    altHoldReset();
+}
+
+void altHoldProcessTransitions(void) {
+
+    if (FLIGHT_MODE(ALTHOLD_MODE)) {
+        if (!altHoldState.isAltHoldActive && isAltitudeAvailable()) {
+            altHoldReset();
+            altHoldState.isAltHoldActive = true;
+        }
+    } else {
+        altHoldState.isAltHoldActive = false;
+    }
+
+    // ** the transition out of alt hold (exiting altHold) may be rough.  Some notes... **
+    // The original PR had a gradual transition from hold throttle to pilot control throttle
+    // using !(altHoldRequested && altHoldState.isAltHoldActive) to run an exit function
+    // a cross-fade factor was sent to mixer.c based on time since the flight mode request was terminated
+    // it was removed primarily to simplify this PR
+
+    // hence in this PR's the user's throttle needs to be close to the hover throttle value on exiting altHold
+    // its not so bad because the 'target adjustment' by throttle requires that
+    // user throttle must be not more than half way out from hover for a stable hold
+}
+
+void altHoldUpdateTargetAltitude(void)
+{
+    // The user van raise or lower the target altitude with throttle up;  there is a big deadband.
+    // Max rate for climb/descend is 1m/s by default (up to 2.5 is allowed but overshoots a fair bit)
+    // If set to zero, the throttle has no effect.
+
+    // Some people may not like throttle being able to change the target altitude, because:
+    // - with throttle adjustment, hitting the switch won't always hold altitude if throttle is bumped
+    // - eg if the throttle is bumped low accidentally, quad will start descending.
+    // On the plus side:
+    // - the pilot has control nice control over altitude, and the deadband is wide
+    // - Slow controlled descents are possible, eg for landing
+    // - fine-tuning height is possible, eg if there is slow sensor drift
+    // - to keep the craft stable, throttle must be in the deadband, making exits smoother
+
+    const float rcThrottle = rcCommand[THROTTLE];
+
+    const float lowThreshold = 0.5f * (positionConfig()->hover_throttle + PWM_RANGE_MIN); // around 1150
+    const float highThreshold = 0.5f * (positionConfig()->hover_throttle + PWM_RANGE_MAX); // around 1500
+    
+    float throttleAdjustmentFactor = 0.0f;
+    if (rcThrottle < lowThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, PWM_RANGE_MIN, lowThreshold, -1.0f, 0.0f);
+    } else if (rcThrottle > highThreshold) {
+        throttleAdjustmentFactor = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+    }
+
+    // if failsafe is active, and we get here, we are in landing mode, so descend at configured max rate
```

Comment:
This override is a bit confusing - code around landing mode may change, this implicit condition os creating hidden dependency.
Maybe even consider passing argument that will specify desired altitude update (throttle/landing/<some future mode>

Reply:
Yes there are three situations where we control throttle:
- landing mode
- gps rescue
- altitude hold 
It makes sense to make altitude control a function that can be sent a target altitude by each of these clients.
This could be done in mixer.c maybe, because that's where throttle gets mixed in.
I'll try to get the gps rescue altitude code to be as much identical to this code as the first step, but in a different PR.